### PR TITLE
Add missing documentation for X509_ATTRIBUTE related functions.

### PR DESCRIPTION
**Pull request by slontis**
Partial fix for #8026

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: doc/man3/CMS_signed_get_attr.pod
Comment (jamuir) [inline doc/man3/CMS_signed_get_attr.pod:74]: is the reference to `X509_REQ_get_attr_by_OBJ()` correct?  or should we instead reference a `CMS_signed_get` function?

Comment (slontis) [reply doc/man3/CMS_signed_get_attr.pod:74]: ta.. Block copy - I need to re-read everything..

**Comment by t8m** [comment]
I think this should go into all branches.

File: doc/man3/CMS_signed_get_attr.pod
Comment (mattcaswell) [inline doc/man3/CMS_signed_get_attr.pod:69]: Should we add something like "The CMS_unsigned_XXX() functions work in the same way as the CMS_signed_XXX() equivalents".

File: doc/man3/CMS_signed_get_attr.pod
Comment (mattcaswell) [inline doc/man3/CMS_signed_get_attr.pod:106]: Remove "int" from the start of this sentence

File: doc/man3/CMS_signed_get_attr.pod
Comment (mattcaswell) [inline doc/man3/CMS_signed_get_attr.pod:116]: Typo: if there are

File: doc/man3/CMS_signed_get_attr.pod
Comment (mattcaswell) [inline doc/man3/CMS_signed_get_attr.pod:117]: Nit: too many spaces between signed and attribute

File: doc/man3/CMS_signed_get_attr.pod
Comment (mattcaswell) [inline doc/man3/CMS_signed_get_attr.pod:140]: Typo: returns (also add a space before this word)

File: doc/man3/CMS_signed_get_attr.pod
Comment (mattcaswell) [inline doc/man3/CMS_signed_get_attr.pod:190]: s/See/see/ (twice)

File: doc/man3/EVP_PKEY_get_attr.pod
Comment (mattcaswell) [inline doc/man3/EVP_PKEY_get_attr.pod:85]: Nit: s/a/an/

File: doc/man3/X509_ATTRIBUTE.pod
Comment (mattcaswell) [inline doc/man3/X509_ATTRIBUTE.pod:141]: s/a/an/

File: doc/man3/X509_REQ_get_attr.pod
Comment (mattcaswell) [inline doc/man3/X509_REQ_get_attr.pod:95]: ASN.1

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Nice work!

**Comment by slontis** [comment]
Thanks for the review @mattcaswell

**Comment by slontis** [comment]
ping

File: doc/man3/X509_REQ_get_attr.pod
Comment (jamuir) [inline doc/man3/X509_REQ_get_attr.pod:95]: typo:  "ANSN.1" -> "ASN.1"

File: doc/man3/EVP_PKEY_get_attr.pod
Comment (jamuir) [inline doc/man3/EVP_PKEY_get_attr.pod:40]: should we say "used in" rather than "used on"?

File: doc/man3/CMS_signed_get_attr.pod
Comment (jamuir) [inline doc/man3/CMS_signed_get_attr.pod:179]: change `= item` to `=item`

File: doc/man3/EVP_PKEY_get_attr.pod
Comment (jamuir) [inline doc/man3/EVP_PKEY_get_attr.pod:63]: should this be possessive?  "object's"?

File: doc/man3/EVP_PKEY_get_attr.pod
Comment (jamuir) [inline doc/man3/EVP_PKEY_get_attr.pod:96]: should this be possessive?  "object's"?

File: doc/man3/EVP_PKEY_get_attr.pod
Comment (jamuir) [inline doc/man3/EVP_PKEY_get_attr.pod:102]: nit: drop comma.

File: doc/man3/X509_ATTRIBUTE.pod
Comment (jamuir) [inline doc/man3/X509_ATTRIBUTE.pod:76]: maybe use comma (or semi colon), rather than period:  ", i.e." rather than ". i.e."

File: doc/man3/X509_ATTRIBUTE.pod
Comment (jamuir) [inline doc/man3/X509_ATTRIBUTE.pod:85]: This sentence is indented.  I am not sure if it should be.

File: doc/man3/X509_ATTRIBUTE.pod
Comment (jamuir) [inline doc/man3/X509_ATTRIBUTE.pod:187]: `ANS1_TYPE` -> `ASN1_TYPE`

File: doc/man3/X509_ATTRIBUTE.pod
Comment (jamuir) [inline doc/man3/X509_ATTRIBUTE.pod:194]: `ANS1_TYPE` -> `ANS1_TYPE`

File: doc/man3/CMS_signed_get_attr.pod
Comment (jamuir) [inline doc/man3/CMS_signed_get_attr.pod:116]: `ANS1_TYPE` -> `ASN1_TYPE`

File: doc/man3/X509_ATTRIBUTE.pod
Comment (jamuir) [inline doc/man3/X509_ATTRIBUTE.pod:150]: `ANS1_STRING` -> `ANS1_STRING`

File: doc/man3/X509_ATTRIBUTE.pod
Comment (jamuir) [inline doc/man3/X509_ATTRIBUTE.pod:166]: typo: "a issue" -> "an issue"

File: doc/man3/X509_ATTRIBUTE.pod
Comment (jamuir) [inline doc/man3/X509_ATTRIBUTE.pod:182]: nit: missing period at end of sentence.

File: doc/man3/X509_ATTRIBUTE.pod
Comment (jamuir) [inline doc/man3/X509_ATTRIBUTE.pod:205]: nit: missing period.

File: doc/man3/X509_ATTRIBUTE.pod
Comment (jamuir) [inline doc/man3/X509_ATTRIBUTE.pod:213]: maybe use "an" rather than "a"; i.e. `an B<X509_ATTRIBUTE>`

File: doc/man3/CMS_signed_get_attr.pod
Comment (jamuir) [inline doc/man3/CMS_signed_get_attr.pod:109]: typo: "except that is" -> "except that it"

File: doc/man3/X509_REQ_get_attr.pod
Comment (jamuir) [inline doc/man3/X509_REQ_get_attr.pod:67]: typo: "except that is" -> "except that it"

**Review by jamuir — CHANGES_REQUESTED** [review]
that took a while to read through :-)

please see my various comments (all minor).

Thank-you for taking the time to add these missing docs.

File: doc/man3/EVP_PKEY_get_attr.pod
Comment (slontis) [reply doc/man3/EVP_PKEY_get_attr.pod:40]: I think it is ok.

File: doc/man3/CMS_signed_get_attr.pod
Comment (slontis) [reply doc/man3/CMS_signed_get_attr.pod:179]: Well spotted.

**Review by jamuir — APPROVED** [review]

**Comment by mattcaswell** [comment]
Ping @openssl/committers for second review

File: doc/man3/CMS_signed_get_attr.pod
Comment (tom-cosgrove-arm) [inline doc/man3/CMS_signed_get_attr.pod:127]: typo: `similar`

Comment (tom-cosgrove-arm) [reply doc/man3/CMS_signed_get_attr.pod:127]: Also, either

```
CMS_unsigned_XXX() functions are similar to the matching CMS_signed_XXX() functions.
```

or

```
The return values of the CMS_unsigned_XXX() functions are similar to the return
values of the matching CMS_signed_XXX() functions.
```

Comment (tom-cosgrove-arm) [reply doc/man3/CMS_signed_get_attr.pod:127]: But maybe best would be something different like

```
The CMS_unsigned_XXX() functions return values similar to those of the equivalent CMS_signed_XXX() functions.
```

Comment (slontis) [reply doc/man3/CMS_signed_get_attr.pod:127]: I was trying to keep the function name first, but this reads better.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Comment by tom-cosgrove-arm** [comment]
@mattcaswell Do you need/want to reconfirm?

File: doc/man3/CMS_signed_get_attr.pod
Comment (mattcaswell) [inline doc/man3/CMS_signed_get_attr.pod:127]: Nit: I would say "are similar".

**Review by mattcaswell — APPROVED** [review]
Approved with or without the nit fix

File: doc/man3/CMS_signed_get_attr.pod
Comment (tom-cosgrove-arm) [reply doc/man3/CMS_signed_get_attr.pod:127]: Yes, this is better

Comment (jamuir) [reply doc/man3/CMS_signed_get_attr.pod:127]: I think it is fine as is.

Maybe you are both reading it as "their return-values" rather than "they return values".

1. "The CMS_unsigned_XXX() functions return values similar to..."
2. "The CMS_unsigned_XXX() functions' return-values are similar to..." (note the added apostrophe)

Comment (slontis) [reply doc/man3/CMS_signed_get_attr.pod]: Changed it since multiple people thought it was better the other way.

File: doc/man3/CMS_signed_get_attr.pod
Comment (t8m) [inline doc/man3/CMS_signed_get_attr.pod:96]: This is a problem - this behavior is just a recent regression as described in #22266 and we should not ossify it by documenting it. We should NOT reject duplicate attributes being added. I am not sure if there are any signed attributes that can be possibly duplicate but it is definitely true for unsigned attributes.

**Review by t8m — CHANGES_REQUESTED** [review]

File: doc/man3/CMS_signed_get_attr.pod
Comment (slontis) [reply doc/man3/CMS_signed_get_attr.pod:96]: This PR will update depending on what is decided in the other PR. I could only document what it does, not what we would like it to do :)

Comment (t8m) [reply doc/man3/CMS_signed_get_attr.pod:96]: Once we document it so, it becomes ossified as changing it back will be an API change.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@tom-cosgrove-arm @mattcaswell please reconfirm

File: doc/man3/X509_ATTRIBUTE.pod
Comment (tom-cosgrove-arm) [inline doc/man3/X509_ATTRIBUTE.pod:58]: (minor) We seem to have `atrname` here (and also written that way in the text, which is how I noticed it) but `attrname` elsewhere.  It's an inconsistency, but since the text matches the prototypes, it's not technically wrong

**Review by tom-cosgrove-arm — APPROVED** [review]
Reconfirming

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master, 3.2, 3.1 and 3.0 branches. Thank you for your contribution.

**Pull request closed without merging**
